common/quotas/ownership_scaled_quota.go

```diff
@@ -0,0 +1,161 @@
+
+	ErrNegativeConstantScaleFactor = errors.New("constantScaleFactor cannot be negative")
+	ErrNonPositiveTotalNumShards   = errors.New("totalNumShards must be greater than 0")
+)
+
+// Rate returns the rate of the base rate limiter multiplied by the shard ownership share and the constant scale factor.
+func (q *OwnershipScaledQuota) Rate() float64 {
+	return q.rb.Rate() * q.scaleFactor()
+}
+
+// Burst returns the burst quota of the base rate limiter multiplied by the shard ownership share and the constant scale
+// factor, rounded down to the nearest integer.
+func (q *OwnershipScaledQuota) Burst() int {
+	return int(float64(q.rb.Burst()) * q.scaleFactor())
+}
+
+// scaleFactor returns the fraction of the total shards in the cluster owned by this host multiplied by the constant
+// scale factor. It returns 1.0 if there haven't been any shard count updates yet.
+func (q *OwnershipScaledQuota) scaleFactor() float64 {
+	shardCount := q.shardCount.Load()
+	if shardCount == shardCountNotSet {
+		// If the shard count is -1, then we haven't received the first shard count update yet. In this case, we return
+		// 1.0 so that the base quota is not scaled.
+		return 1.0
+	}
+
+	return float64(shardCount) / float64(q.totalShards) * q.constantScaleFactor
+}
+
+// NewOwnershipScaledQuota returns a OwnershipScaledQuota that scales the RPS and burst quotas of the base RateBurst
+// by the fraction of the total shards in the cluster owned by this host. The updateAppliedCallback field is a channel
+// which is sent to in a non-blocking fashion when the shard count updates are applied. This is useful for testing. In
+// production, you should probably pass in nil. If constantScaleFactor is negative or totalNumShards is non-positive,
+// then an error is returned.
+func NewOwnershipScaledQuota(
+	rb RateBurst,
+	shardCounter ShardCounter,
+	constantScaleFactor float64,
+	totalNumShards int,
+	updateAppliedCallback chan struct{},
+) (*OwnershipScaledQuota, error) {
+	if constantScaleFactor < 0 {
+		return nil, fmt.Errorf("%w: %f", ErrNegativeConstantScaleFactor, constantScaleFactor)
+	}
+	if totalNumShards <= 0 {
+		return nil, fmt.Errorf("%w: %d", ErrNonPositiveTotalNumShards, totalNumShards)
+	}
+	subscription := shardCounter.Subscribe()
+	quota := &OwnershipScaledQuota{
+		rb:                    rb,
+		totalShards:           totalNumShards,
+		constantScaleFactor:   constantScaleFactor,
+		subscription:          subscription,
+		updateAppliedCallback: updateAppliedCallback,
+	}
+	// Initialize the shard count to -1 so that we don't apply the scale factor until we receive the first shard count.
+	quota.shardCount.Store(shardCountNotSet)
+	quota.wg.Add(1)
+
+	go quota.startScaling()
+
+	return quota, nil
+}
+
+func (q *OwnershipScaledQuota) startScaling() {
+	defer q.wg.Done()
+
+	for shardCount := range q.subscription.ShardCount() {
+		q.shardCount.Store(int64(shardCount))
+		select {
+		case q.updateAppliedCallback <- struct{}{}:
+		default:
+		}
```

Comment:
probably need to do a nil check on the callback channel.

Reply:
Technically, this works without a nil check because of the default branch and the fact that the nil channel is a valid, always-blocking channel. However, I changed this to be a nil check because it simplifies the code, and I don't know if there's a performance hit for the previous approach.